configserver-client/src/main/java/ai/vespa/hosted/client/AbstractConfigServerClient.java

```diff
@@ -0,0 +1,263 @@
+                    thrown.addSuppressed(e.getCause());
+
+                if (builder.entity != null && ! builder.entity.isRepeatable()) {
+                    log.log(WARNING, "Cannot retry " + request + " as entity is not repeatable");
+                    break;
+                }
+                log.log(FINE, request + " failed; will retry", e.getCause());
+            }
+        }
+        if (thrown != null) {
+            if (thrown instanceof IOException)
+                throw new UncheckedIOException((IOException) thrown);
+            else if (thrown instanceof RuntimeException)
+                throw (RuntimeException) thrown;
+            else
+                throw new IllegalStateException("Illegal retry cause: " + thrown.getClass(), thrown);
+        }
+
+        throw new IllegalArgumentException("No hosts to perform the request against");
+    }
+
+    /** Append path to the given host, which may already contain a root path. */
+    static URI concat(URI host, URIBuilder pathAndQuery) {
+        URIBuilder builder = new URIBuilder(host);
+        List<String> pathSegments = new ArrayList<>(builder.getPathSegments());
+        if ( ! pathSegments.isEmpty() && pathSegments.get(pathSegments.size() - 1).isEmpty())
+            pathSegments.remove(pathSegments.size() - 1);
+        pathSegments.addAll(pathAndQuery.getPathSegments());
+        try {
+            return builder.setPathSegments(pathSegments)
+                    .setParameters(pathAndQuery.getQueryParams())
+                    .build();
+        }
+        catch (URISyntaxException e) {
+            throw new IllegalArgumentException("URISyntaxException should not be possible here", e);
+        }
+    }
+
+    @Override
+    public RequestBuilder send(HostStrategy hosts, Method method) {
+        return new RequestBuilder(hosts, method);
+    }
+
+    /** Builder for a request against a given set of hosts. */
+    class RequestBuilder implements ConfigServerClient.RequestBuilder {
+
+        private final Method method;
+        private final HostStrategy hosts;
+        private final URIBuilder uriBuilder = new URIBuilder();
+        private HttpEntity entity;
+        private RequestConfig config = defaultRequestConfig;
+
+        private RequestBuilder(HostStrategy hosts, Method method) {
+            if ( ! hosts.iterator().hasNext())
+                throw new IllegalArgumentException("Host strategy cannot be empty");
+
+            this.hosts = hosts;
+            this.method = requireNonNull(method);
+        }
+
+        @Override
+        public RequestBuilder at(String... pathSegments) {
+            uriBuilder.setPathSegments(requireNonNull(pathSegments));
+            return this;
+        }
+
+        @Override
+        public ConfigServerClient.RequestBuilder body(byte[] json) {
+            return body(HttpEntities.create(json, ContentType.APPLICATION_JSON));
+        }
+
+        @Override
+        public RequestBuilder body(HttpEntity entity) {
```

Comment:
Consider adding body overloads for Slime object and Jackson type as well.

Reply:
Yes, I considered this back and forth, and so far I think that's better placed in the users of this, but I may still change my mind. 
```java
    .map(bytes -> mapper.readValue(bytes, Some.class)
```
and 
```java
    .map(someSerializer::fromJson)
```
seem like the best approach right now, imho.

Reply:
I guess it was called `read(...)` now, not `map(...)`, but whatever.